### DISCUSSION TITLE
isa-debug-exit port/device?

### DISCUSSION THREAD

**Discussion by posk-io** [cloud-hypervisor/cloud-hypervisor#3484] [Q&A] [answered]
Hello!

In qemu, one can specify 

``` 
-device isa-debug-exit,iobase=0xf4,iosize=0x04
```

and in the guest kernel/firmware, one can use

```
pub fn exit_qemu(exit_code: QemuExitCode) {
    use x86_64::instructions::port::Port;

    unsafe {
        let mut port = Port::new(0xf4);
        port.write(exit_code as u32);
    }
}
````

(see https://os.phil-opp.com/testing/ for details and use case/motivation). 

Is there a similar feature in cloud-hypervisor, i.e. an easy way to terminate/exit from inside the guest? If not, would you be open to adding it? (I can send a patch/PR, but it may be not up to your standards/style/pracitices).

Thanks!

**Comment by rbradford** [comment, marked as answer]
I would suggest using the `SLEEP_CONTROL_REG` which is exposed as PIO 0x3c0 (https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/vmm/src/acpi.rs#L205). You should write the value 0x34 to trigger the shutdown: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/devices/src/acpi.rs#L51

This is how the firmware & Linux kernel trigger the VM to exit.

**Reply by posk-io** [reply]
That works - thanks, Rob!
